### PR TITLE
[Frontend][TFLite] Add parser support for shape and range

### PR DESCRIPTION
**Pull request by dhruvaray**
@inadob , @siju-samuel, @anijain2305 kindly review

### CODE REVIEW

File: python/tvm/relay/frontend/tflite.py [hunk 1]
```diff
@@ -579,6 +581,49 @@ def convert_tanh(self, op):
...
+        assert isinstance(op, Operator)
+        input_tensors = self.get_input_tensors(op)
+        assert len(input_tensors) == 3, "input tensors length should be 3"
+
+        start, limit, delta = self.get_expr(input_tensors[0].tensor_idx),\
+                              self.get_expr(input_tensors[1].tensor_idx),\
+                              self.get_expr(input_tensors[2].tensor_idx)
+
+
+        out = _op.arange(start, limit, delta, "int32")
```
Comment (siju-samuel) [inline python/tvm/relay/frontend/tflite.py:604]: Why only integer range support?

File: tests/python/frontend/tflite/test_forward.py [hunk 2]
```diff
@@ -650,6 +693,51 @@ def test_all_resize():
         _test_resize(tf.image.resize_nearest_neighbor, data, align_corners=False)
 
 
+#######################################################################
+# Range
+# -------------
+def test_range():
```
Comment (siju-samuel) [inline tests/python/frontend/tflite/test_forward.py:699]: test_range and test_shape is not invoked from __main__

File: tests/python/frontend/tflite/test_forward.py [hunk 3]
```diff
@@ -650,6 +693,51 @@ def test_all_resize():
...
+                ["start", "limit", "delta"],
+                [start,limit,delta],
+                [out],
+                mode="vm",
+                quantized=False
+        )
+#######################################################################
+# Shape
+# -------------
+def test_shape():
```
Comment (siju-samuel) [inline tests/python/frontend/tflite/test_forward.py:720]: keep the naming convention same as other cases, test_forward_shape(), same for range also.

File: tests/python/frontend/tflite/test_forward.py [hunk 4]
```diff
@@ -650,6 +693,51 @@ def test_all_resize():
...
+                [x for x in np.nditer(data)],
+                ["start", "limit", "delta"],
+                [start,limit,delta],
+                [out],
+                mode="vm",
+                quantized=False
+        )
+#######################################################################
+# Shape
+# -------------
```
Comment (siju-samuel) [inline tests/python/frontend/tflite/test_forward.py:719]: Have more dashes than needed

File: tests/python/frontend/tflite/test_forward.py (hunk 2 above)
Comment (dhruvaray) [reply tests/python/frontend/tflite/test_forward.py]: added...

File: tests/python/frontend/tflite/test_forward.py (hunk 3 above)
Comment (dhruvaray) [reply tests/python/frontend/tflite/test_forward.py]: Fixed...

File: tests/python/frontend/tflite/test_forward.py (hunk 4 above)
Comment (dhruvaray) [reply tests/python/frontend/tflite/test_forward.py]: Fixed...

File: python/tvm/relay/frontend/tflite.py (hunk 1 above)
Comment (dhruvaray) [reply python/tvm/relay/frontend/tflite.py]: @siju-samuel : Fixed and added extra test cases to demonstrate support for other types. I inter the type though...

**Comment by dhruvaray** [comment]
@siju-samuel - Please confirm if you are ok with the changes so that this can be merged...

File: python/tvm/relay/frontend/tflite.py [hunk 5]
```diff
@@ -579,6 +582,63 @@ def convert_tanh(self, op):
...
+        """Convert TFLite Shape"""
+        try:
+            from tflite.Operator import Operator
+        except ImportError:
+            raise ImportError("The tflite package must be installed")
+
+        if self.is_quantized(op):
+            raise tvm.error.OpNotImplemented(
+                'TFlite quantized SHAPE operator is not supported yet.')
+
```
Comment (siju-samuel) [inline python/tvm/relay/frontend/tflite.py:633]: Does the shape output have any impact with quantized inputs? do we need this check?

File: tests/python/frontend/tflite/test_forward.py [hunk 6]
```diff
@@ -188,7 +231,7 @@ def compare_tflite_with_tvm(in_data, in_name, input_tensors,
                 continue
 
             tvm_output = run_tvm_graph(tflite_model_buffer, in_data, in_node, target=device,
-                                       num_output=len(out_names), out_names=out_names)
+                                       num_output=len(out_names), out_names=out_names,mode=mode)
```
Comment (siju-samuel) [inline tests/python/frontend/tflite/test_forward.py:234]: `out_names,mode` -> space after comma

File: tests/python/frontend/tflite/test_forward.py [hunk 7]
```diff
@@ -650,6 +693,82 @@ def test_all_resize():
...
+            out = tf.shape(r, out_type=tf.dtypes.int32)
+            compare_tflite_with_tvm(
+                [x for x in np.nditer(data)],
+                ["start", "limit", "delta"],
+                [start, limit, delta],
+                [out],
+                mode="vm",
+                quantized=False
+            )
 #######################################################################
```
Comment (siju-samuel) [inline tests/python/frontend/tflite/test_forward.py:772]: add new lines here

File: python/tvm/relay/frontend/tflite.py [hunk 8]
```diff
@@ -579,6 +582,63 @@ def convert_tanh(self, op):
...
+        for t in [start, limit, delta]:
+            if self.has_expr(t.tensor_idx):
+                expressions.append(self.get_expr(t.tensor_idx))
+            else:
+                tensor_type = self.get_tensor_type_str(t.tensor.Type())
+                tensor_value = self.get_tensor_value(t)
+                expressions.append(self.exp_tab.new_const(tensor_value, dtype=tensor_type))
+
+        #out type inference
```
Comment (siju-samuel) [inline python/tvm/relay/frontend/tflite.py:612]: #out  ->  # out .......
Add space before starting comment. Change in all places

File: python/tvm/relay/frontend/tflite.py [hunk 9]
```diff
@@ -579,6 +582,63 @@ def convert_tanh(self, op):
...
+        expressions = []
+
+        for t in [start, limit, delta]:
+            if self.has_expr(t.tensor_idx):
+                expressions.append(self.get_expr(t.tensor_idx))
+            else:
+                tensor_type = self.get_tensor_type_str(t.tensor.Type())
+                tensor_value = self.get_tensor_value(t)
+                expressions.append(self.exp_tab.new_const(tensor_value, dtype=tensor_type))
+
```
Comment (siju-samuel) [inline python/tvm/relay/frontend/tflite.py:611]: use `get_tensor_expr` function from recent prs.

File: python/tvm/relay/frontend/tflite.py [hunk 10]
```diff
@@ -579,6 +582,63 @@ def convert_tanh(self, op):
 
         return out
 
+    def convert_range(self, op):
+        """Convert TFLite Range"""
+        try:
+            from tflite.Operator import Operator
```
[the remark this answers is not in the document]
Comment (dhruvaray) [reply python/tvm/relay/frontend/tflite.py]: fixed...

File: tests/python/frontend/tflite/test_forward.py (hunk 6 above)
Comment (dhruvaray) [reply tests/python/frontend/tflite/test_forward.py]: fixed...

File: python/tvm/relay/frontend/tflite.py (hunk 8 above)
Comment (dhruvaray) [reply python/tvm/relay/frontend/tflite.py]: fixed...

File: python/tvm/relay/frontend/tflite.py (hunk 9 above)
Comment (dhruvaray) [reply python/tvm/relay/frontend/tflite.py]: Yes... fixed...

File: tests/python/frontend/tflite/test_forward.py (hunk 7 above)
Comment (dhruvaray) [reply tests/python/frontend/tflite/test_forward.py]: added

File: python/tvm/relay/frontend/tflite.py (hunk 5 above)
Comment (dhruvaray) [reply python/tvm/relay/frontend/tflite.py]: removed these checks...

**Comment by dhruvaray** [comment]
@siju-samuel - rebased and incorporated comments. Kindly review.

**Comment by dhruvaray** [comment]
@siju-samuel - Let me know if all is good

**Comment by dhruvaray** [comment]
@siju-samuel : Range returns dynamic tensors and graph does not support that...

**Comment by dhruvaray** [comment]
@FrozenGene - Could you kindly review and merge?

**Comment by dhruvaray** [comment]
@anijain2305 - Please review and merge

**Comment by anijain2305** [comment]
@siju-samuel Can you please approve the PR if you are ok with the changes?
https://tvm.apache.org/docs/contribute/code_review.html#approve-and-request-changes-explicitly

**Comment by anijain2305** [comment]
Thanks @dhruvaray @siju-samuel This is merged!

**Pull request merged**